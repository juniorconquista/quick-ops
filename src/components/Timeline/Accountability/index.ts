import Accountability from './accountability';

export default Accountability;
