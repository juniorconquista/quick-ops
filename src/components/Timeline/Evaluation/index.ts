import Evaluation from './evaluation';

export default Evaluation;
