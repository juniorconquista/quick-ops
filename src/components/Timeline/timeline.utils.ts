import moment from 'moment';

import { ReactComponent as Hotel } from '../../assets/icons/hotel-bell.svg';
import { ReactComponent as Food } from '../../assets/icons/burger.svg';
import { ReactComponent as Users } from '../../assets/icons/group.svg';
import { ReactComponent as Asterisk } from '../../assets/icons/asterisk.svg';

export const handleIcon = (type: string) => {
    switch (type) {
        case 'concierge-bell':
            return Hotel;
        case 'utensils':
            return Food;
        case 'evaluation':
            return Users;
        default:
            return Asterisk;
    }
};

export const formatDate = (date: number) => moment(date).format('DD/MM/YYYY');
