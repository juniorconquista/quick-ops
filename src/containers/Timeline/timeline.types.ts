export enum CARDTYPE {
    EXPENSE = 'EXPENSE',
    EVALUATION = 'EVALUATION',
    ACCOUNTABILITY_SUBMITTED = 'ACCOUNTABILITY_SUBMITTED',
}

export type Card = {
    author?: {
        name: string;
    };
    amountSpent: string;
    amountTotal: string;
    cardDate: any;
    cardType: CARDTYPE;
    contentType:
        | 'hotel-fee'
        | 'EVALUATION'
        | 'ACCOUNTABILITY_SUBMITTED'
        | 'ACCOUNTABILITY_CREATED';
    currencySymbol: string;
    // expenseEvaluation?: number;
    // expenseId: number;
    expenseTypeCode:
        | 'hotel-fee'
        | 'EVALUATION'
        | 'ACCOUNTABILITY_SUBMITTED'
        | 'ACCOUNTABILITY_CREATED';
    expenseTypeIcon: string;
    // expenseTypeId: number;
    id: number;
    // invoiceDate: number;
    notes: string;
    // resourceUrl: string;
    status?: 'base' | 'PENDING' | 'APPROVED';
    // updatedOn: number;
};

export interface TimelineProps {
    content: Card[];
}
