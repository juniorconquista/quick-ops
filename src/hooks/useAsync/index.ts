import useAsync from './useAsync';
export * from './useAsync.utils';

export default useAsync;
