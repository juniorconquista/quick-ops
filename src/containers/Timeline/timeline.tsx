import React, { memo, useEffect } from 'react';
import { Box } from 'quarks-ui';

import handlerFactory from './timeline.handleFactory';
import { CARDTYPE, TimelineProps } from './timeline.types';
import useAsync, { headers } from '../../hooks/useAsync';
import Loading from '../../components/Loading';

const Timeline: React.FC = () => {
    const getTimeline = (): Promise<TimelineProps> =>
        fetch(`${process.env.REACT_APP_API_URL}/timeline`, {
            headers,
        }).then((response) => response.json());

    const { execute, value, status } = useAsync<TimelineProps>(
        getTimeline,
        false,
    );

    useEffect(() => {
        execute();
    }, []); // eslint-disable-line react-hooks/exhaustive-deps

    return (
        <Box styling="column" width="100%" mt="24px">
            {value?.content?.map((card) => {
                const CardComponent = handlerFactory(card.cardType as CARDTYPE);
                return <CardComponent key={card.id} {...card} />;
            })}
            {status === 'pending' && (
                <Box styling="row" height="250px" position="relative">
                    <Loading />
                </Box>
            )}
        </Box>
    );
};

export default memo(Timeline);
