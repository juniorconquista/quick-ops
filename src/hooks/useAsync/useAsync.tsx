import { useState, useEffect, useCallback } from 'react';
import { Status } from './useAsync.types';

const useAsync = <T, E = string>(
    asyncFunction: (body?: any) => Promise<T>,
    immediate = true,
) => {
    const [status, setStatus] = useState<Status>('idle');
    const [value, setValue] = useState<T | null>(null);
    const [error, setError] = useState<E | null>(null);

    const execute = useCallback(
        (body?: any) => {
            setStatus('pending');
            setValue(null);
            setError(null);

            return asyncFunction(body)
                .then((response: any) => {
                    setValue(response);
                    setStatus('success');
                })
                .catch((error: any) => {
                    setError(error);
                    setStatus('error');
                });
        },
        [asyncFunction],
    );

    useEffect(() => {
        if (immediate) {
            execute();
        }
    }, [execute, immediate]);

    return { execute, status, value, error };
};

export default useAsync;
