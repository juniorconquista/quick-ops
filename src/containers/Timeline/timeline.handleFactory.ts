import Expense from '../../components/Timeline/Expense';
import Evaluation from '../../components/Timeline/Evaluation';
import Accountability from '../../components/Timeline/Accountability';

import { CARDTYPE } from './timeline.types';

const handler = (type: CARDTYPE) => {
    switch (type) {
        case CARDTYPE.EXPENSE:
            return Expense;
        case CARDTYPE.EVALUATION:
            return Evaluation;
        default:
            return Accountability;
    }
};

const handlerFactory = (type: CARDTYPE) => handler(type);
export default handlerFactory;
