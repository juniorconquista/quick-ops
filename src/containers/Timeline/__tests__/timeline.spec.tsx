import React from 'react';
import { cleanup, render } from '@testing-library/react';

import { TimelineFaker } from '../../../mocks/timeline.faker';
import * as useAsync from '../../../hooks/useAsync';
import ThemeProvider from '../../../utils/ThemeProviderTest';
import Timeline from '../timeline';
import 'jest-styled-components';

const setupFetchStub: any = (data: any) => (_url: string) =>
    new Promise((resolve) => {
        resolve({
            json: () =>
                Promise.resolve({
                    data,
                }),
        });
    });

const timelineResponseFaker = new TimelineFaker();
const timelineResponse = timelineResponseFaker.generate();

describe('Timeline container', () => {
    let mockUseAsyncSpyFn: any;

    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });

    beforeEach(() => {
        // spyOn's
        const fakeData = { fake: 'data' };
        jest.spyOn(global, 'fetch').mockImplementation(
            setupFetchStub(fakeData),
        );

        const useAsyncSpy = jest.spyOn(useAsync, 'default');
        mockUseAsyncSpyFn = {
            ...jest.requireActual('../../../hooks/useAsync'),
            execute: jest.fn(),
            value: timelineResponse,
        };
        useAsyncSpy.mockImplementation(
            jest.fn((fn) => {
                fn();
                return mockUseAsyncSpyFn;
            }),
        );
    });

    it('tests basic snapshot', () => {
        const useAsyncSpy = jest.spyOn(useAsync, 'default');
        const mockUseAsyncSpyFn = {
            ...jest.requireActual('../../../hooks/useAsync'),
            execute: jest.fn(),
            value: {},
        };
        useAsyncSpy.mockImplementation(
            jest.fn((fn) => {
                fn();
                return mockUseAsyncSpyFn;
            }),
        );

        const container = render(
            <ThemeProvider>
                <Timeline />
            </ThemeProvider>,
        );
        expect(container.asFragment()).toMatchSnapshot();
    });

    it('should call the api to bring the necessary data when rendering the timeline', () => {
        render(
            <ThemeProvider>
                <Timeline />
            </ThemeProvider>,
        );
        expect(mockUseAsyncSpyFn.execute).toBeCalled();
    });
});
