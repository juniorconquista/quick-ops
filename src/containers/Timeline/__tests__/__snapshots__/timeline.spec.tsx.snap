// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline container tests basic snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 24px;
  width: 100%;
}

.c0:focus {
  outline: none;
}

<div
    class="c0"
    width="100%"
  />
</DocumentFragment>
`;
